### PR TITLE
telem_view Report FPS and Batch UI Updates By Measurement Group

### PR DESCRIPTION
**Pull request by AarC10**
# Description
telem_view was using an obscene amount (enough to make my laptop fans audible when monitoring backplane over Ethernet). This is because UI updates were called for each measurement instead of by group instead of being batched by group. 

 Added an FPS counter to see how often the UI is being updated and updates are batched by measurement group now. Argument for limiting updates too if a user chooses to with the -fps flag.

# How Has This Been Tested?
Compared running the old and new version of telem_view. Dropped ~850 updates a second down to ~95 updates per second when monitoring backplane. Monitored CPU usage which dropped from ~9% down to ~0.5%.

# Checklist:
- [x] New functionality is documented in the necessary spots (i.e new functions documented in the header)
- [x] Unit tests cover any new functionality or edge cases that the PR was meant to resolve (if applicable) 
- [x] The CI checks are passing
- [x] I reviewed my own code in the GitHub diff and am sure that each change is intentional

### CODE REVIEW

File: cmd/telem_view/telem_view.go [hunk 1]
```diff
@@ -110,7 +135,22 @@ func main() {
...
 
 				offset := 0
+
+				// prep slices to collect all updates for this packet
+				countMeas := len(pkt.Measurements)
+				valStrs := make([]string, countMeas)
+				hexStrs := make([]string, countMeas)
+				binStrs := make([]string, countMeas)
+
+				// capture flags locally tso we avoid closure/capture races
```
Comment (K3das) [inline cmd/telem_view/telem_view.go]:
```suggestion
				// capture flags locally so we avoid closure/capture races
```

File: cmd/telem_view/telem_view.go [hunk 2]
```diff
@@ -110,7 +135,22 @@ func main() {
...
+				hexLocal := hexOn
+				binLocal := binOn
+
 				for i, name := range pkt.Measurements {
+					valStrs[i] = padValue("–")
+					hexStrs[i] = ""
+					binStrs[i] = ""
```
Comment (K3das) [inline cmd/telem_view/telem_view.go]: I believe this is unnecessary because the slice is already zero-initialized. Not sure if go is smart enough to optimize this.

File: cmd/telem_view/telem_view.go [hunk 3]
```diff
@@ -18,6 +20,10 @@ import (
 const valueColWidth = 12
 
 var shmDir = flag.String("shm", "/dev/shm", "directory to use for shared memory")
+var fpsLimit = flag.Int("fps", 0, "max UI frames per second (0 = unlimited)")
+
+var updateCounter uint64
+var pendingUpdate int32
```
Comment (K3das) [inline cmd/telem_view/telem_view.go]: Could these be their atomic equivalents? They're shared across threads and all operations to them are already atomic.

File: cmd/telem_view/telem_view.go [hunk 4]
```diff
@@ -80,16 +95,51 @@ func main() {
...
+	}()
+
+	go func() {
+		var interval time.Duration
+		if *fpsLimit > 0 {
+			interval = time.Second / time.Duration(*fpsLimit)
+		} else {
+			// when fps limit is 0 (unlimited), do ~60 FPS
+			// 1000 / 60 = 16.67ms
+			interval = 17 * time.Millisecond
```
Comment (K3das) [inline cmd/telem_view/telem_view.go]: Nit-pick, but could this be done with `time.Second / 60`?

**Review by K3das — APPROVED** [review]
yippee
